### DISCUSSION TITLE
High level project organization: submodules, plug-ins (plug-in managers), project subdivision, etc...

### DISCUSSION THREAD

**Discussion by charliewhitfield** [ivoyager/forum#7] [General]
This is an open discussion for anyone. But I'll start with this...

Quoting @Difint:
> I've spend a while looking into code, and here are my thoughts:
> 
> 1. There is a nice plugin manager for Godot [gd-plug](https://github.com/imjp94/gd-plug), combined with git submodules, allows quite flawless decomposition of the project into several libraries
> 2. Here are my thoughts about the possible decomposition:
> 
> * File loading and IO operations like tsv/csv parsing
> * GUI
> * Visual rendering of solar system/planets
> * Simulation core
> * Assets/Resource dowloader (new feature, highly needed, I'm planning to implement soon, happy to discuss)
> 
> Such a decomposition would be helpful for me, particularly File IO & Visual part.
> 
> Right now I'm working on a detailed 3D map of the Moon based on NASA CGI map, and I'm stuck with the task of detailed rendering of planets. As a simulation core I'm planning to make a GDextension for GMAT and rely on GMAT features, as that code has TRL-9.
> 
> To summarize, I'm Happy to collaborate on those tasks.

I'm totally favorable to splitting off parts of `ivoyager` into separate repositories, with the idea that they could stand alone in the Godot Asset Library. Our *.tsv table reader is a good first candidate (it's quite powerful, if I'm allowed to say so). I guess I also see the idea of splitting our GUI widgets into a library, although this seems like a lower priority (they are useless without `ivoyager`, but I guess someone might want `ivoyager` without its GUI widgets).

But first, let me ask some questions about using a plug-in manager. These aren't meant to sound negative. I'm just trying to cover my own knowledge gaps:

1. We already make use of git submodules. In fact, I'm building my own game from several repositories (public and private) just using git submodules. What can a plugin manager do that I can't do with submodules? OK, I see that gd-plug can clean out my .import directory. I get the impression that it might be easier to add/remove add-ons on the fly, so to speak. What else can a plug-in manager do for me? (Note: I do hear that some folks dislike or struggle with git submodules. I find them reasonably easy to use, possibly because I use GitKraken.)
2. Can an external project make use of the ivoyager submodule (and/or associated submodules) using their own plug-in manager, even if I, Voyager organization doesn't include or specify use of any manager? 2b. If this were to be our approach, is there anything we can do to make that easier?
3. [Possibly related to your assets/resource downloader???] We have our `ivoyager_assets` directory which, conceptually, should just be another submodule. However, I have to ask folks to download it rather than maintaining it as a public repository, which is awkward. Can gd-plug help with this? (FYI, I maintain `ivoyager_assets` as a private git-lfs repository, but distribute it only as a download with `ivoyager` releases. The reason is that GitHub seems happy to host my downloads for free, but will charge me storage/bandwidth if someone forks my git-lfs repository. Also, I don't want to require users to have git-lfs. Also, sometimes git-lfs seems a little flaky/unstable, at least in conjunction with GitKraken.)

**Comment by 0xIonRod** [comment]
Sorry for the late reply

- Git submodules are very convinient during the development process, when you actually modify the submodule and commit to git. However, it introduces certain overheads managing git, like additional commands, or manuel update of the submodule if you forked it.
- A plugin called gd-plug is a very convinient tool to manage dependencies. It can handle custom folder structure, to support current ivoyager's structure.
- Same pain here. gd-plug can't handle it, so I'm using submodule + git-lfs, thou I'm planning to create a plugin that would handle automatic assets download from a URL

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Comment by charliewhitfield** [comment]
The decomposition discussion and the plugin manager discussion are orthogonal. We can decompose `ivoyager` into separate repositories with or without a plugin manager.

I'm kind of ignorant about plugin managers (and package managers, is this even related?). In my understanding, we have three options:
1. Make gd-plug integral and required in any I, Voyager project. By required I mean we allow code structures that can't easily be maintained with git submodules alone.
2. Include gd-plug in `ProjectTemplate` but in a non-required, non-integral way. I.e., a developer can simply remove it or replace it with another manager.
3. Don't include a plugin manager. Developers can add a manager into their own project directory, or not, as they choose.

I don't like option 1. It seems to me that option 3 is exactly where we are right now.

**Comment by charliewhitfield** [comment]
In principle, our .tsv table loader/reader system should be an editor plugin. Each table would then be an imported Resource with it's own read API.

Pros:
- More Godoty.
- Faster app startup. (It's tiny though: the Planetarium .tsv files are parsed in only ~90 ms on my laptop.)
- Smaller exported app. (This is also small: the files add up to 66 kb.)
- I think this is the right approach if we want to distribute this system in the Godot Asset Library.

Cons:
- Devs would have to add the editor plugin.

@Difint, does gd-plug help with Godot Editor plugins?

This is not high on my priority list (and I haven't made or worked with editor plugins yet). Just thinking out loud.

**Reply by 0xIonRod** [reply]
Hi, sorry for slow replies.

1. Regarding gd-plug - for sure it should be possible to develop without it. However, it's really nice and convinient, and helps a lot with all Addons, including Editor plugin. Mostly, I was talking about decomposing according to Godot suggestion - all libs should be stored in the /addons folder
2. It'll be really convenient with resources, and it's relatively easy to make an importer, however from my perspective the main benifit would be a possibility for a wider community to use TSV/CSV importer separately, without the ivoyager. It could help to make the code better. Also it'll allow to use other TSV/CVS importer libs

**Reply by charliewhitfield** [reply]
Slow is good. Gives me time to code... ;-)

I think my line of questioning is not very clear. I'm not asking whether or why a plugin manager is useful. I'm asking: What should we (the I, Voyager Organization) do about that?

1. Facilitate use but don't include a plugin manager in our own repositories. "Facilitate" includes project decomposition and directory restructuring (and possibly other things I'm not aware of). Developers can choose whether or not they want to add a plugin manager to their own project.
2. Encourage use. Same as above but we include gd-plug in our [project_template](https://github.com/ivoyager/project_template) repository. A developer could still remove it or replace it with another manager if they want.

What I don't want to do is restructure in some way that **requires use**. I.e., developers can't develop without gd-plug. I'm still reluctant on my own game project, as git submodules give me extremely precise version control. To be specific, my under-development game project is often pointing to a non-release ivoyager commit. (But this is my own choice. It shouldn't limit other devs.)

**Comment by 0xIonRod** [comment]
Hi, sorry for being not clear. Form my perspective It would be very useful if:

- You could move the module into addons/ folder as Godot suggests
- Organize all modules like regular projects, storing the code in the addons/ folder of the module git repository
- Maybe decompose the ivoyager core into several submodules like ivoyager-loaders, ivoyager-visualsation, ivoyager-orbital-mechanics and so on.

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Reply by charliewhitfield** [reply]
I think my confusion is that I didn't understand that you can run two Godot **"projects"** at the same time. However, I guess that is exactly what you are doing with multiple editor plugins. (Seriously, I only thought of plugins as editor enhancement stuff. Not actual project content.) I think I will go and read the plugin docs now to clear up my ignorance...

So, in the big picture, "I, Voyager" would simply be a set of editor plugins. Is that the idea?

**Comment by charliewhitfield** [comment]
[Edit:] TL;DR (long post):
Godot doc's recommendation for plugin distribution requires one to manually move directories. This is simply not acceptable for code maintenance. Instead, we'll be distributing our plugins so that the _entire plugin directory_ goes inside the addons directory. The user can still do manual directory manipulations (if they choose) but have the option to add and maintain the code as a git submodule:

`git submodule add https://github.com/ivoyager/some_plugin addons/some_plugin `

.
.
.
OK, I've finally read up on Godot addons. I'm trying to get my head around **Godot project** versus **GitHub repository** structures. I'm spelling this out in great detail because I'm literally horrified by version control difficulty if I try to do this as recommended. (But perhaps I'm missing something.)

For this discussion I want to talk about two actual Godot projects: our public "Planetarium" and my own (private) in-development game "MyGame" (it's Astropolis but let's keep discussion generic).

Let's further assume that we've split I, Voyager into three parts, "ivoyager_tables", "ivoyager_core" and "ivoyager_gui", and we want to structure them correctly as addons.

Here are the actual working **project** directory structures:

```
planetarium
  |-addons
  |  |- ivoyager_tables
  |  |- ivoyager_core
  |  |- ivoyager_gui
  |-planetarium_content
  |-project.godot

my_game
  |-addons
  |  |- ivoyager_tables
  |  |- ivoyager_core
  |  |- ivoyager_gui
  |-my_game_content
  |-project.godot
```
Now ... when I look at the above structures ... the thing that makes sense to me is to make `ivoyager_tables`, `ivoyager_core` and `ivoyager_gui` into GitHub submodules (in other words, three GitHub repositories that are not Godot projects). This makes version control quite easy using my present work flow. To give an example:

1. While working on the Planetarium, I make a change in `planetarium/addons/ivoyager_core/`.
2. I push above change to the `ivoyager_core` submodule.
3. Next time I'm working on MyGame, I pull updated `ivoyager_core` submodule which updates `my_game/addons/ivoyager_core/`.
4. While working on MyGame, I make a change in `my_game/addons/ivoyager_core/`.
5. I push above change to the `ivoyager_core` submodule.
6. Next time I'm working on Planetarium, I pull the updated `ivoyager_core` submodule which updates `planetarium/addons/ivoyager_core/`.

Different dir structure, but I've done exactly the above a thousand times or more while working on the Planetarium and my game project simultaneously over the last five years.

The important concept here is that I'm never working on the submodule (or plugin) code in isolation. It's **always** in a real project when making or testing code changes.

However, if I'm reading correctly, the Godot plugin docs seem to want the plugins to be distributed as "full-fledged projects" as follows:

```
ivoyager_tables_plugin
 |- addons
 |  |- ivoyager_tables
 |      |- plugin.cfg
 |      |- ivoyager_tables.gd
 |- project.godot

ivoyager_core_plugin
 |- addons
 |  |- ivoyager_core
 |      |- plugin.cfg
 |      |- ivoyager_core.gd
 |- project.godot

ivoyager_gui_plugin
 |- addons
 |  |- ivoyager_gui
 |      |- plugin.cfg
 |      |- ivoyager_gui.gd
 |- project.godot
```
Honestly, I'm not happy about code maintenance in this structure. How on God's Green Earth am I supposed to do version control!!!??? (Pardon the expletive.)

To be specific about the problem (and sorry for being repetative): I'm never ever working in `ivoyager_core_plugin/addons/ivoyager_core/`. Instead, I'm working in `planetarium/addons/ivoyager_core/` or `my_game/addons/ivoyager_core/`. Am I supposed to cut/paste my changes back to the plugin repositories? That's insane!

.
.
OK, after some thought...

~One way I can make this work is to now have **eight** GitHub repositories:~
* ~planetarium~
* ~my_game~
* ~ivoyager_tables_plugin~
* ~ivoyager_core_plugin~
* ~ivoyager_gui_plugin~
* ~ivoyager_tables (submodule)~
* ~ivoyager_core (submodule)~
* ~ivoyager_gui (submodule)~

~So all 8 above are GitHub repositories, but only the first 5 are Godot projects (of which 3 are plugins). The 3 plugin repositories each point to their respective submodule. In this scenario, plugin users would have to use `--recursive` when cloning a plugin. It's doable but a bit unwieldy.~
[Edit: No, not only is that a dumb idea, but Godot's docs say not to have submodules in your addon.]

.
.
Some more thought...

Well, I guess I could do cut/paste code maintenance on plugins if I **only** update plugin repositories for major releases. So I keep submodules for my own work flow. To maintain the plugin repository I (quite literally) cut/paste code from my submodule to the plugin repository.

But seriously, the above doesn't work. What if someone sends a pull request on one of the plugins? I'd have to cut/paste the change back to my working submodules to test it. This is not workable!

**Comment by 0xIonRod** [comment]
Thx for the detailed explanation! I'be being cracking my head about project structure for a while as well.

Personally I've stuck to the architecture that you've called "Here are the actual workingprojectdirectory structures:"

Sorry for the mess, but now things are cleared and I'm really happy with the structure you've proposed.

I came to the similar conclusin, and folder structure, and I've got confused by the Godot suggested structure.

The only missing point is how to make assets management easier. I think it's related to this issue, as I'm using a separate repository for binaries with git-lfs, but it's super inconvenient.

It feels like there is no module, that could download binary assets, so I'm thinking of doing that

Sent with [Proton Mail](https://proton.me/) secure email.

------- Original Message -------

**Reply by charliewhitfield** [reply]
[Edit: deleted my annoying response] Let's just go with a structure we can maintain. Hopefully users will be ok with that.

**Comment by charliewhitfield** [comment]
After breaking out the table importer, I'm thinking this will be our future project structure (using planetarium as example parent project):

```
planetarium
 |- addons
     |- ivoyager_core
          |- ivoyager_assets
     |- ivoyager_table_importer
 |- planetarium_stuff
```
I still have mixed feelings about splitting out ivoyager's GUI widgets and popups. They don't do anything unless the user actually adds them in their own GUI. (Well, except for taking up namespace with a lot of 'IVSomeButtonWidget' names...)

[Edit:] I'm going back and forth on GUI split, and where to put the assets. Maybe:

```
planetarium
 |- addons
     |- ivoyager_assets
     |- ivoyager_core
     |- ivoyager_gui
     |- ivoyager_table_importer
 |- planetarium_stuff
```

**Comment by 0xIonRod** [comment]
By moving GUI to separate addon it'll allow others (me specifically) to use this as a "windows" library and not only build on top of it, but also add new features

1. Also it'll help to make other parts of the code clear e.g. move static/widgets.gd from ivoyager core to the gui repo
2. Overall files from static/ folder could be moved to a separate "utils" repo as they could be reused in other projects

**Comment by charliewhitfield** [comment]
[Edit:] This is not workable now that I've learned that `add_custom_type()` doesn't make real classes and appears to be depreciated or on the way to being depreciated...

~The GUI widgets may be a moot point due to the way I'm reworking [ivoyager_core](https://github.com/ivoyager/ivoyager_core) as an editor plugin. Here is the new system I'm working on:~

- ~There will be no 'class_name' lines anywhere in ivoyager_core.~
- ~All classes are defined in 'ivoyager.cfg' and 'ivoyager_override.cfg'. The latter file is inited at the project directory level if it doesn't exist already. These config files are used by the plugin to add all ivoyager classes as custom editor classes.~
- ~In 'ivoyager_override.cfg' you can remove, add, replace or subclass individual classes defined in the core 'ivoyager.cfg'.~
- ~In 'ivoyager_override.cfg' we also have convenience settings you can set to remove whole systems:~

~[classes]~

~remove_all_gui_widgets=false~
~remove_all_gui_popups=false~
~remove_all_save_load_system=false~
~remove_all_camera_system=false~

- ~In 'ivoyager_override.cfg' you can also change project settings that are defined in singleton IVGlobal.~
- ~In 'ivoyager_override.cfg' you can even change the autoload singletons (e.g., you can replace IVGlobal).~

~So the only cost of leaving the GUI widgets in core is the size of the files on disk, which currently adds up to a whopping 142 kb. If user sets 'remove_all_gui_widgets=true' then these are never loaded and there is no name-space clutter.~

~I see your point about separating static utility classes that might be useful outside of ivoyager. However, the GUI widgets are entirely dependent on ivoyager code: they have no use whatsoever without ivoyager_core.~

~This new class system has the added benefit of not using `.godot/global_script_class_cache.cfg`, which is currently buggy and doesn't update for changes made outside of the editor (e.g., when updating a submodule).~

~(This system will replace our existing 'project_builder.gd', which had some of this functionality but couldn't change the global class namespace.)~

**Reply by charliewhitfield** [reply]
Well, never mind I guess...

It appears that `add_custom_type()` is more or less being depreciated. It doesn't make real classes that you can use in any real way, e.g., MyClass.new() doesn't work.

I also didn't realize that classes defined with `class_name` are supposed to NOT be available when the file is in addons and the plugin is inactivated. I don't know if that is documented anywhere but I've seen post comments to that effect. So that argues for more plugin decomposition.

**Comment by 0xIonRod** [comment]
Thx for the clarification about the GUI.

I'm looking for a Windows manager, so probably I'll go with [Panku Console ](https://github.com/Ark2000/PankuConsole) in the common/lynx_window2


Also Panku console has an interesting Design Solution when they use "modules".


I'll be lookind deep into IVoyager code and Panku console in the next days and will provide feedback with ideas

**Comment by 0xIonRod** [comment]
A few more thoughts:
1. The extension system is cool, thou it's not obvious how it's used, especially initialization. Thou it would be cool to have it as an addon  to allow to reuse in other projects
2. Same for windows system, maybe not to move GUI, but just window management system

I'm still learning how the system works, and how to integrate with it, as now everything is managed by the IVoyager core (e.g. initialization). That has it's own pros and cons, and I need more time to think of it

**Comment by charliewhitfield** [comment]
Take a look at [ivoyager_core](https://github.com/ivoyager/ivoyager_core). It will replace 'ivoyager'. It has no sim code yet but it does have the new initialization system I'm working on. It will let you init ivoyager in a way similar to the existing system (modifying values in our singletons at init) or to do so entirely via an 'ivoyager.cfg' config file, or some mix of the two.

The existing IVProjectBuilder was designed with the idea that developers would start with our system and build a game or app from it. In the new paradigm I'm trying to make it easier to go the other way: i.e., start with an existing project and then add ivoyager. So it won't force you to make a specific 'extension' file in a specific way. You'll just add ivoyager plugins to any project and modify it however you want.

**Reply by 0xIonRod** [reply]
The best news on Sunday morning!

Yep, the new approach would help to integrate with IVoyager!

**Comment by charliewhitfield** [comment]
Regarding "window management system". I've tried to encapsulate this sort of thing in classes like the existing 'window_manager.gd'. These already can be removed from the project build by modifying a dictionary item in IVProjectBuilder. In the new system you can do that either by modifying a dictionary item in IVInitializer (similar to current system) or by specifying dictionary changes via 'ivoyager.cfg'.

**Comment by 0xIonRod** [comment]
Does it makes sense to move si_base_units to IVoyager core?

**Comment by charliewhitfield** [comment]
It won't be needed in the new system because the EditorPlugin will add IVUnits as an autoload singleton, and it will define its own base units. And the user can specify a replacement path for that using 'ivoyager.cfg' in the project directory.

It has to be a singleton or static file so we can use constants in other files: e.g., 'const SOME_DISTANCE = 2.0 * IVUnits.AU'. It was in the project directory so user could change base units (METER in particular for scaling) without modifying the submodule code. But with the editor plugin we can set singletons dynamically, so problem solved...

**Comment by 0xIonRod** [comment]
There was a post with the roadmap and idea to check latest assets automatically.

It looks like this feature is implemented in [PankuConsole](https://github.com/Ark2000/PankuConsole/tree/master/addons/panku_console/modules/check_latest_release) (as well as a lot of other interesting things, like Windows management and modules loading system)

**Comment by charliewhitfield** [comment]
I built an assets loader! It's working in the current master branch.

I wanted something very specific. The way it works is that ivoyager_core editor plugin knows exactly what assets version it needs and where to download it, even if it is not a release version. It prompts the user at editor start (or plugin enabling) to 'Download' or 'Cancel' if assets are missing or there is a version mismatch. So... if you revert back to an earlier version of ivoyager_core, it will prompt you to download the corresponding (earlier) assets version (which you may or may not want to do).

In the future we'll have an option to 'Update to Latest'. But for now, the link to the exact dev version is more helpful to me.
